gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -294,49 +330,145 @@ public void run() {
                                 UInt32Value.newBuilder().setValue(objectHasher.hash().asInt())));
                   }
                 }
-                requestObserver.onNext(requestBuilder.build());
-
-                if (objectFinalized) {
-                  // Close the request half of the streaming RPC.
-                  requestObserver.onCompleted();
+                if (dataChunkMap.size() == numberOfRequestsToRetain) {
+                  dataChunkMap.remove(dataChunkMap.firstKey());
+                  retryMap.remove(dataChunkMap.firstKey());
                 }
+                dataChunkMap.put(writeOffset, chunkData);
+                return requestBuilder;
               }
 
               private ByteString readRequestData() throws IOException {
-                // Mark the input stream in case this request fails so that read can be recovered
-                // from where it's marked.
-                pipeSource.mark(MAX_BYTES_PER_MESSAGE);
                 ByteString data =
                     ByteString.readFrom(ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE));
-
-                objectFinalized =
-                    data.size() < MAX_BYTES_PER_MESSAGE || pipeSource.available() <= 0;
+                pipeSource.mark(1);
+                objectFinalized = data.size() < MAX_BYTES_PER_MESSAGE || pipeSource.read() == -1;
+                pipeSource.reset();
                 return data;
               }
+
+              // Handles the case when a writeOffset of data read previously is being processed.
+              // This happens if a transient failure happens while uploading, and can be resumed by
+              // querying the current committed offset.
+              private InsertObjectRequest.Builder buildRequestFromCachedDataChunk(
+                  TreeMap<Long, ByteString> dataChunkMap) {
+                // Resume will only work if the first request builder in the cache carries an offset
+                // not greater than the current writeOffset.
+                InsertObjectRequest.Builder requestBuilder = null;
+                if (dataChunkMap.size() > 0 && dataChunkMap.firstKey() <= writeOffset) {
+                  long prevWriteOffset = 0;
+                  ByteString prevDataChunk = null;
+                  for (Map.Entry<Long, ByteString> entry : dataChunkMap.entrySet()) {
```

Comment:
Because you already use TreeSet, may you use binary search methods to find offset that you need instead of iterating over them?

Reply:
The list is only 5 long, so binary search is less efficient than just walking the list linearly.
I also previously suggested a TreeSet is overkill - all we need here is a simple LinkedList. Henry didn't want to do that because he wants to get the offset from the tree key but if you instead use as LinkedList of request messages (rather than just the byte strings) you have the offsets in the request messages and it has the other important advantage that you avoid re-scanning the full chunk data whenever re-transmitting.

Again, I think it's time for a meeting. I feel the current implementation is both overly complex and inefficient.

Reply:
Discussed offline. Will keep TreeSet and plain iteration for the sake of simplicity and readability.